src/osd/ECTransaction.h

```diff
@@ -120,6 +132,28 @@ class ECTransaction : public PGBackend::PGTransaction {
     assert(len == bl.length());
     ops.push_back(AppendOp(hoid, off, bl, fadvise_flags));
   }
+  void write(
+    const hobject_t &hoid,
+    uint64_t off,
+    uint64_t len,
+    bufferlist &bl,
+    uint32_t fadvise_flags) {
+    if (len == 0) {
+        touch(hoid);
+        return;
+    }
+    offset_write = true;
+
+    written += off + len > written ? (off + len - written) : 0;
+    assert(len == bl.length());
+    writeops.push_back(WriteOp(hoid, off, len, bl, fadvise_flags));
+  }
+  WriteOp* get_writeop() {
+    assert(!writeops.empty());
+    WriteOp* op = &(writeops.front());
```

Comment:
It seems single item is enough, I am not sure.

2016-02-03 21:49 GMT+08:00 Igor Fedotov <notifications@github.com>:

> In src/osd/ECTransaction.h
> <https://github.com/ceph/ceph/pull/7481#discussion_r51721993>:
>
> > +    uint64_t len,
> > +    bufferlist &bl,
> > +    uint32_t fadvise_flags) {
> > +    if (len == 0) {
> > +        touch(hoid);
> > +        return;
> > +    }
> > +    offset_write = true;
> > +
> > +    written += off + len > written ? (off + len - written) : 0;
> > +    assert(len == bl.length());
> > +    writeops.push_back(WriteOp(hoid, off, len, bl, fadvise_flags));
> > +  }
> > +  WriteOp* get_writeop() {
> > +    assert(!writeops.empty());
> > +    WriteOp* op = &(writeops.front());
>
> As you never "pop_front" from writeops - this is either incorrect or you
> don't need a container (single item is enough)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/7481/files#r51721993>.
>